sources/api/netdog/src/networkd/mod.rs

```diff
@@ -1,6 +1,88 @@
-mod config;
+pub(crate) mod config;
+mod conversions;
 mod devices;
 
+use self::config::{NetDevConfig, NetworkConfig, NetworkDConfigFile};
+use self::devices::{NetworkDDevice, NetworkDInterface};
+use crate::interface_id::{InterfaceId, InterfaceName};
+use std::collections::HashMap;
+
+// A map of network device -> associated VLANs.  This type exists to assist in generating a
+// device's network configuration, which must contain it's associated VLANs.
+pub(self) type Vlans = HashMap<InterfaceName, Vec<InterfaceName>>;
+
+pub(crate) struct NetworkDConfig {
+    devices: Vec<NetworkDDevice>,
+    vlans: Vlans,
+}
+
+impl NetworkDConfig {
+    pub(crate) fn new<T>(network_devices: Vec<T>) -> Result<Self>
+    where
+        T: TryInto<NetworkDDevice, Error = self::error::Error>,
+    {
+        let mut devices = network_devices
+            .into_iter()
+            .map(|d| d.try_into())
+            .collect::<Result<Vec<NetworkDDevice>>>()?;
+
+        let mut device_names = Vec::with_capacity(devices.len());
+        let mut vlans = HashMap::new();
+        for device in &devices {
+            device_names.push(device.name());
+            // VLANs are typically attached to a network device (bond/interface). Net config
+            // specifies the bond or interface in the VLAN config. In systemd-networkd config, the
+            // VLAN is specified in the device config. Create a map of interface/bond name to list
+            // of attached VLANs to assist with config generation later
+            if let NetworkDDevice::Vlan(vlan) = device {
+                vlans
+                    .entry(vlan.device.clone())
+                    .or_insert_with(Vec::new)
+                    .push(vlan.name.clone())
+            }
+        }
+
+        // If the VLANs map contains a device we don't otherwise have config for, it means the
+        // device is used only as a link for the VLAN.  This is used in VLAN "tagged only" type
+        // setups.  Add an empty NetworkDInterface to the list of config to be generated.  An empty
+        // device will get a .network file so it is managed and becomes a memeber of the VLAN, but
+        // will otherwise have all DHCP and addressing config turned off.
+        for vlan_device in vlans.keys() {
+            if !device_names.contains(&InterfaceId::from(vlan_device.clone())) {
```

Comment:
I'm fine either way, but figured I should bring this up now since it occurred to me while reading it. Do we plan to use `device_names` anywhere else later? This logic modified `devices` which is what generates `device_names` but the set isn't re-generated. I wonder if it's worth adding newly created devices to this `Vec` just in case future changes want to reuse this, otherwise we have a potential gotcha for those following afterwards.

Reply:
It _is_ possible that `device_names` could be useful later, but I'd rather not add any additional loops to this code to regenerate a list we don't currently use.